### ISSUE TITLE
consume.ReceiveAsync() use CancellationToken  may cause a deadlock.

### ISSUE THREAD

**Issue by vbirds** [fsharplang-ru/pulsar-client-dotnet#224]
conseumer ReceiveAsync() with timeout CancellationToken  may cause a deadlock.  The following code runs after a while，will stuck at var message = await consumer.ReceiveAsync(cts.Token); and cannot return .

version: nutgut 2.12.2
framework: net7.0
build type: release

```c#
using Pulsar.Client.Api;
using Pulsar.Client.Common;

class Program
{
    static async Task Main()
    {
        string url = "pulsar://192.168.1.42:6650";
        
        PulsarClient? client = await new PulsarClientBuilder()
            .ServiceUrl(url)
            .BuildAsync();

        IConsumer<byte[]> consumer = await client.NewConsumer(Schema.BYTES())
            .Topic("persistent://public/test2/test")
            .SubscriptionName("test_group")
            .ConsumerName("test")
            .SubscriptionType(Pulsar.Client.Common.SubscriptionType.Shared)
            .SubscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
            .AckTimeout(TimeSpan.FromSeconds(20))
            .SubscribeAsync();

        while (true)
        {
            try
            {
                using (var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(2)))
                {
                    Console.WriteLine("{0} test xx begin",  DateTime.Now);
                    var message = await consumer.ReceiveAsync(cts.Token);
                    Console.WriteLine("{0} test xx end",  DateTime.Now);
                    if (message != null)
                    {
                        await consumer.AcknowledgeAsync(message.MessageId);
                        Console.WriteLine("{0} recv a message",  DateTime.Now);
                    }
                    
                }
            }
            catch (OperationCanceledException)
            {
                Console.WriteLine("{0} test xx empty",  DateTime.Now);
            }
            catch (Exception e)
            {
                Console.WriteLine("{0} [Error] ReceiveAsync error:{1}",  DateTime.Now, e.ToString());
                break;
            }
        }

        await client.CloseAsync();
    }
}
```

**Comment by Lanayx** [comment]
Hi! How long is "for a while"? I've run this example and it was working for more than a minute without issues, then I stopped it.

**Comment by vbirds** [comment]
> Hi! How long is "for a while"? I've run this example and it was working for more than a minute without issues, then I stopped it.

Hello @Lanayx
Thank you for your reply.
This deadlock phenomenon occurs after the program has been running for 15 minutes.

**Comment by fjod** [comment]
Run it for 10 minutes on windows10 + pulsar in docker, no issues

**Comment by vbirds** [comment]
> Run it for 10 minutes on windows10 + pulsar in docker, no issues

It appeared after I had been running it for 15 minutes. on windows10 + pulsar in docker,

**Comment by fjod** [comment]
> > Run it for 10 minutes on windows10 + pulsar in docker, no issues
> 
> It appeared after I had been running it for 15 minutes. on windows10 + pulsar in docker,

run it for about 1 hour, saw only prints like this:

```
03.07.2023 10:59:19 test xx empty
03.07.2023 10:59:19 test xx begin
03.07.2023 10:59:19 test xx empty
03.07.2023 10:59:19 test xx begin
03.07.2023 10:59:19 test xx empty
03.07.2023 10:59:19 test xx begin
```

**Comment by vbirds** [comment]
> > > Run it for 10 minutes on windows10 + pulsar in docker, no issues
> > 
> > 
> > It appeared after I had been running it for 15 minutes. on windows10 + pulsar in docker,
> 
> run it for about 1 hour, saw only prints like this:
> 
> ```
> 03.07.2023 10:59:19 test xx empty
> 03.07.2023 10:59:19 test xx begin
> 03.07.2023 10:59:19 test xx empty
> 03.07.2023 10:59:19 test xx begin
> 03.07.2023 10:59:19 test xx empty
> 03.07.2023 10:59:19 test xx begin
> ```

 i try running it for 15 minutes. it whill stuck at var message = await consumer.ReceiveAsync(cts.Token);  and cannot return .

**Comment by Lanayx** [comment]
@vbirds Since you are the only one who can reproduce the issue, you'd want to start investigating it yourself. For example, try collecting relevant client logs, checking memory consumption, taking memory dump, try using source code to debug the issue instead of nuget etc

**Comment by vbirds** [comment]
> @vbirds Since you are the only one who can reproduce the issue, you'd want to start investigating it yourself. For example, try collecting relevant client logs, checking memory consumption, taking memory dump, try using source code to debug the issue instead of nuget etc

Thank you for your guide. I whill try it.

**Comment by Lanayx** [comment]
I've just published 2.12.3 version, there is a small chance that it will resolve your issue as well, would you please check

**Comment by vbirds** [comment]
> I've just published 2.12.3 version, there is a small chance that it will resolve your issue as well, would you please check

Thank you for your reply. i whill check it.

**Comment by Lanayx** [comment]
Closing as abandoned
